runtime/lua/vim/lsp/semantic_tokens.lua

```diff
@@ -0,0 +1,577 @@
+  local state = state_by_client[ctx.client_id]
+
+  if not client or not state or err then
+    return vim.NIL
+  end
+
+  for bufnr, _ in pairs(state.buffers) do
+    reset_buf_state(bufnr, client, true)
+
+    for _, win in ipairs(api.nvim_list_wins()) do
+      if bufnr == api.nvim_win_get_buf(win) then
+        request_all_tokens(bufnr, ctx.client_id)
+        break
+      end
+    end
+  end
+
+  return vim.NIL
+end
+
+--- on_win handler for the decoration provider (see |nvim_set_decoration_provider|)
+---
+---If there is a current result for the buffer and the version matches the
+---current document version, then the tokens are valid and can be applied. As
+---the buffer is drawn, this function will add extmark highlights for every
+---token in the range of visible lines. Once a highlight has been added, it
+---sticks around until the document changes and there's a new set of matching
+---highlight tokens available.
+---
+---If this is the first time a buffer is being drawn with a new set of
+---highlights for the current document version, the namespace is cleared to
+---remove extmarks from the last version. It's done here instead of the response
+---handler to avoid the "blink" that occurs due to the timing between the
+---response handler and the actual redraw.
+---
+---@private
+local function do_win_highlight(_, _, bufnr, topline, botline)
+  if not active_buffers[bufnr] then
+    return false
+  end
+
+  for _, client in pairs(vim.lsp.buf_get_clients(bufnr)) do
+    local state, buf_state = get_state(bufnr, client.id)
+
+    if state and buf_state and buf_state.current_result.version == util.buf_versions[bufnr] then
+      local current_result = buf_state.current_result
+      if not current_result.namespace_cleared then
+        api.nvim_buf_clear_namespace(bufnr, state.namespace, 0, -1)
+        current_result.namespace_cleared = true
+      end
+
+      -- We can't use ephemeral extmarks because the buffer updates are not in
+      -- sync with the list of semantic tokens. There's a delay between the
+      -- buffer changing and when the LSP server can respond with updated
+      -- tokens, and we don't want to "blink" the token highlights while
+      -- updates are in flight, and we don't want to use stale tokens because
+      -- they likely won't line up right with the actual buffer.
+      --
+      -- Instead, we have to use normal extmarks that can attach to locations
+      -- in the buffer and are persisted between redraws.
+      local highlights = current_result.highlights
+      local idx = binary_search(highlights, topline)
+
+      for i = idx, #highlights do
+        local token = highlights[i]
+
+        if token.line > botline then
+          break
+        end
+
+        if not token.extmark_added then
+          api.nvim_buf_set_extmark(
+            bufnr,
+            state.namespace,
+            token.line,
+            token.start_col, {
+              hl_group = '@' .. token.type,
```

Comment:
`@` were introduced for tree-sitter with some magic to support fallback with scopes. Would it make sense to add a different kind of magic highlight groups that is different from the tree-sitter ones and that might take into account token modifiers. Maybe similar to https://vscode-westeu.azurewebsites.net/api/language-extensions/semantic-highlight-guide#semantic-coloring-in-color-themes 

> `variable.readonly:java` is called a selector and has the form `(*|tokenType)(.tokenModifier)*(:tokenLanguage)?`

this could also be realized using the `@` fallbacks

Reply:
No. We don't want to add yet a third type of highlight names. In particular, we _want_ to use the same highlight groups as for treesitter for tokens of the same name.

(And yes, one way of realizing modifiers similar to how VS Code does it is to support globs in highlight groups like `@foo.*.baz`, but I don't know if that's possible -- @bfredl ?)

Reply:
We talked about this a bit offline. We do want to support modifiers (and potentially language specific groups), but we wanted to wait to add anything more extensible to a future PR after we get initial support in, which will simply link the LSP's token type to a same-named highlight group.

Modifiers might make more sense once we get the ability to do true highlight attribute combining, instead of trying to support an exponential explosion of highlight groups, or until we have something more akin to VSCode's "selector" that supports wildcards, like @clason was referring to.

Reply:
@clason VSCode does not support globs. The star just stands for having arbitrary modifiers in you mapping. E.g. `variable.readonly.global` matches a token of type `variable` with modifiers `readonly` and `global`. When the token is variable and has at least the two modifiers. VScode uses this notation just in configuration tables (or just consider token type for now). Neovim could do the same and then just use built-in highlight groups like `Type`. 

I was just confused why `@` was chosen. `@` from captures and names from LSP are two completely different sources and the same name might have different meanings. I would expect them to be mapped independently form each other. Color schemes would need to consider a union of tree-sitter captures and LSP token types. We could then at least provide a list of that union as the `@` groups that have a default mapping in built-in Nvim. So the planned the default experience with semantic token is that only the groups that have the same name as tree-sitter captures get a default mapping?

Reply:
> @clason VSCode does not support globs. The star just stands for having arbitrary modifiers in you mapping.

I know; what I'm saying is that globs is how _we_ would probably realize that.

> I was just confused why @ was chosen. @ from captures and names from LSP are two completely different sources and the same name might have different meanings.

No. They refer to the same basic structures. More names is just more confusion.

> I would expect them to be mapped independently form each other. 

Why?

> Color schemes would need to consider a union of tree-sitter captures and LSP token types

Yes. Luckily the vast majority of tokens already exist as tree-sitter captures (with the same meaning).

And, again, we're not talking about modifiers yet in this PR.

Reply:
To @theHamsta 's point, it does reduce flexibility by not having an LSP specific group in the mix.

The original way I wanted to do this was what he's talking about--create hlgroups for each semantic token type and default link them to tree sitter highlight groups where it makes sense. Generally the meanings are the same but forcing them together without any possible way to configure them differently seems unnecessarily prohibitive.

Doing this would give the same out of the box experience but gives more flexibility to colorscheme authors to tweak tree sitter and LSP toke s independently. If a colorscheme author doesn't care, the default links would still give semantic tokens meaningful highlights.

However, this is something we can explore in a future PR, as this one doesn't actually introduce any _new_ highlight groups that we may want to break later.

Reply:
We discussed this; it increases maintenance (and documentation) surface for no apparent gain. Let's wait until someone comes up with a compelling(!) actual use case for distinguishing treesitter `@function` and LSP `function`.

Reply:
The most compelling reason is simply that we'll need some links anyway, because not all semantic token type names directly map to a tree sitter capture name. If we need some, then it's actually more maintenance burden (IMO) to determine which token types need mappings because there isn't a tree sitter capture with the same name, and which do. Just create special LSP ones for all of them and map them all to something or nothing.

Again, future PR. Let's leave this one as is for now.

Reply:
> The most compelling reason is simply that we'll need some links anyway, because not all semantic token type names directly map to a tree sitter capture name. 

No, we already have links: `@function` links to `Function`. We'd just add the missing ones to that (even if nvim-treesitter doesn't use them -- yet). The whole point is that we do _not_ want to add another layer of indirection, where LSP highlights map to treesitter highlights map to default (Vim) highlights.

Reply:
I still disagree with this, but we can talk about it more on a future PR.

Reply:
> If you feel very strongly about this, use @function.lsp (which falls back to @function). But again, let's cross that bridge when we get to it.

This was the idea that I had first.  I had the feeling that it was less maintenance to not set nothing at all (or just have a simplistic mapping that can be overwritten). But probably just caring in the default experience for `@XXX` where `xxx` is already handled by tree-sitter mappings is not to bad. It doesn't have to be stable and people see something until someone convinces with something better.